remote_attestation/rust/src/attestation.rs

```diff
@@ -0,0 +1,176 @@
+            let tee_report = AttestationReport::new(&public_key_hash);
+            Some(AttestationInfo {
+                report: Some(tee_report),
+                certificate: tee_certificate.clone(),
+            })
+        } else {
+            None
+        };
+
+        Ok(AttestationIdentity {
+            public_key,
+            attestation_info,
+        })
+    }
+
+    pub fn attest(
+        self,
+        peer_identity: &AttestationIdentity,
+    ) -> anyhow::Result<AttestationStateMachine<AttestedState>> {
+        if let Some(expected_tee_measurement) = &self.state.expected_tee_measurement {
+            // Verify peer attestation info.
+            let peer_attestation_info = peer_identity
+                .attestation_info
+                .as_ref()
+                .context("Peer identity doesn't contain attestation info")?;
+            Self::verify_attestation(&peer_attestation_info, &expected_tee_measurement)
+                .context("Couldn't verify peer attestation info")?;
+        }
+
+        // Generate session key.
+        let session_key = self
+            .state
+            .key_negotiator
+            .derive_session_key(peer_identity.public_key.as_ref())
+            .context("Couldn't derive session key")?;
+
+        let encryptor = AeadEncryptor::new(session_key);
+        Ok(AttestationStateMachine::<AttestedState> {
+            state: AttestedState { encryptor },
+        })
+    }
+
+    /// Verifies the validity of the attestation info:
+    /// - Checks that the TEE report is signed by TEE Provider’s root key.
+    /// - Checks that the public key hash from the TEE report is equal to the hash of the public key
+    ///   presented in the server response.
+    /// - Extracts the TEE measurement from the TEE report and compares it to the
+    ///   `expected_tee_measurement`.
+    fn verify_attestation(
+        peer_attestation_info: &AttestationInfo,
+        expected_tee_measurement: &[u8],
+    ) -> anyhow::Result<()> {
+        // TODO(#1867): Add remote attestation support, use real TEE reports and check that
+        // `AttestationInfo::certificate` is signed by one of the root certificates.
+        peer_attestation_info
+            .verify()
+            .context("Couldn't verify peer attestation info")?;
+
+        // Verify peer tee measurement.
+        let peer_report = peer_attestation_info
+            .report
+            .as_ref()
+            .context("Couldn't find report in peer attestation info")?;
+        if expected_tee_measurement == peer_report.measurement {
+            Ok(())
+        } else {
+            Err(anyhow!("Incorrect TEE measurement"))
+        }
+    }
+}
+
+pub struct AttestedState {
+    /// Utility object for decrypting and encrypting messages using a Diffie-Hellman session key.
+    encryptor: AeadEncryptor,
+}
+
+impl AttestationStateMachine<AttestedState> {
+    pub fn encrypt(&mut self, data: &[u8]) -> anyhow::Result<EncryptedData> {
```

Comment:
Since the return types here and in `decrypt` are not `AttestationStateMachine`, I don't think these functions need to be defined on `AttestationStateMachine`.